### ISSUE TITLE
Loading, Screens, and ui.close()

### ISSUE THREAD

**Issue by shawna-p** [renpy/renpy#3379]
I've been getting an unusual error occasionally when loading a saved file from my game. It can take the form of a few different issues; "spurious ui.close()" and "ui.interact called with non-empty widget/layer stack. Did you forget a ui.close() somewhere?" are the other two. Unfortunately it's been very difficult to pin down exactly what causes the error, as loading the game multiple times seems to almost cycle through them.

My game does have somewhat unusual `after_load` functionality, as I always try to return the player to a particular called screen, regardless of where the save left off (it serves as a "hub" of sorts, from which the player can navigate through small story chunks).

One of the errors I get is the following:

```
I'm sorry, but an uncaught exception occurred.

While running game code:
  File "game/chat_system/chat_setup.rpy", line 239, in script
    call screen loading(0.2, Show('home_screen'))
  File "renpy/common/000statements.rpy", line 569, in execute_call_screen
    store._return = renpy.call_screen(name, *args, **kwargs)
AttributeError: 'NoneType' object has no attribute 'get'

-- Full Traceback ------------------------------------------------------------

Full traceback:
  File "game/chat_system/chat_setup.rpy", line 239, in script
    call screen loading(0.2, Show('home_screen'))
  File "renpy/ast.py", line 2015, in execute
    self.call("execute")
  File "renpy/ast.py", line 2003, in call
    return renpy.statements.call(method, parsed, *args, **kwargs)
  File "renpy/statements.py", line 278, in call
    return method(parsed, *args, **kwargs)
  File "renpy/common/000statements.rpy", line 569, in execute_call_screen
    store._return = renpy.call_screen(name, *args, **kwargs)
  File "renpy/exports.py", line 3136, in call_screen
    rv = renpy.ui.interact(mouse="screen", type="screen", roll_forward=roll_forward)
  File "renpy/ui.py", line 298, in interact
    rv = renpy.game.interface.interact(roll_forward=roll_forward, **kwargs)
  File "renpy/display/core.py", line 3337, in interact
    scene_lists.replace_transient()
  File "renpy/display/core.py", line 945, in replace_transient
    self.remove(layer, tag, prefix=prefix)
  File "renpy/display/core.py", line 1261, in remove
    self.hide_or_replace(layer, remove_index, prefix)
  File "renpy/display/core.py", line 1175, in hide_or_replace
    d = oldsle.displayable._in_current_store()._hide(now - st, now - at, prefix)
  File "renpy/display/screen.py", line 527, in _hide
    hid.child = renpy.ui.default_fixed(focus="_screen_" + "_".join(self.screen_name))
  File "renpy/ui.py", line 514, in __call__
    old_main = screen.old_widgets.get(widget_id, None)
AttributeError: 'NoneType' object has no attribute 'get'

Windows-10-10.0.19041
Ren'Py 7.4.11.2266
Demo 1.0
Fri Apr  1 19:46:03 2022
```

In digging through the code, I did notice the following in `renpy/ui.py`:

```py
        # Should we transfer data from an old version of this screen?
        old_transfers = screen and screen.old_transfers

        # Should we add?
        do_add = True

        if screen:
            if widget_id in screen.widget_properties:
                keyword.update(screen.widget_properties[widget_id])

            if widget_id in screen.hidden_widgets:
                do_add = False

        if old_transfers:
            old_main = screen.old_widgets.get(widget_id, None) # type: ignore

            if self.replaces and old_main is not None:
                keyword["replaces"] = old_main
        else:
            old_main = None
```

Notably, Ren'Py checks for `screen` and `screen.old_transfers` but doesn't actually check if `screen.old_widgets` is None before performing a `get` call, and this seems to be the direct cause of the traceback (though quite likely not the entire issue). By the end of the update function in `screens.py`, it's also possible for `self.old_widgets` to be set to None while `self.old_transfers` is True, which would be a combination of values that would cause the above issue.

Of course, this is mostly speculation on my part, as I haven't thoroughly dug through the logic to understand exactly what's happening and thought it might be quicker to bring it up here.

This is another one of the errors: 
```
I'm sorry, but an uncaught exception occurred.

While running game code:
  File "game/chat_system/chat_setup.rpy", line 241, in script
    call screen home_screen
  File "renpy/common/000statements.rpy", line 569, in execute_call_screen
    store._return = renpy.call_screen(name, *args, **kwargs)
Exception: Spurious ui.close().

-- Full Traceback ------------------------------------------------------------

Full traceback:
  File "game/chat_system/chat_setup.rpy", line 241, in script
    call screen home_screen
  File "renpy/ast.py", line 2015, in execute
    self.call("execute")
  File "renpy/ast.py", line 2003, in call
    return renpy.statements.call(method, parsed, *args, **kwargs)
  File "renpy/statements.py", line 278, in call
    return method(parsed, *args, **kwargs)
  File "renpy/common/000statements.rpy", line 569, in execute_call_screen
    store._return = renpy.call_screen(name, *args, **kwargs)
  File "renpy/exports.py", line 3136, in call_screen
    rv = renpy.ui.interact(mouse="screen", type="screen", roll_forward=roll_forward)
  File "renpy/ui.py", line 298, in interact
    rv = renpy.game.interface.interact(roll_forward=roll_forward, **kwargs)
  File "renpy/display/core.py", line 3325, in interact
    repeat, rv = self.interact_core(preloads=preloads, trans_pause=trans_pause, pause=pause, pause_start=pause_start, **kwargs)
  File "renpy/display/core.py", line 3592, in interact_core
    self.compute_overlay()
  File "renpy/display/core.py", line 2926, in compute_overlay
    renpy.ui.close()
  File "renpy/ui.py", line 384, in close
    stack[-1].close(d)
  File "renpy/sl2/slast.py", line 180, in close
    raise Exception("Spurious ui.close().")
Exception: Spurious ui.close().

Windows-10-10.0.19041
Ren'Py 7.4.11.2266
Demo 1.0
Fri Apr  1 20:03:45 2022
```

And the final alternative error:

```
I'm sorry, but an uncaught exception occurred.

While running game code:
  File "game/chat_system/chat_setup.rpy", line 239, in script
    call screen loading(0.2, Show('home_screen'))
  File "renpy/common/000statements.rpy", line 569, in execute_call_screen
    store._return = renpy.call_screen(name, *args, **kwargs)
Exception: ui.interact called with non-empty widget/layer stack. Did you forget a ui.close() somewhere?
Stack was <Layer: u'transient'>
<renpy.ui.Detached object at 0x0000000007f27950>

-- Full Traceback ------------------------------------------------------------

Full traceback:
  File "game/chat_system/chat_setup.rpy", line 239, in script
    call screen loading(0.2, Show('home_screen'))
  File "renpy/ast.py", line 2015, in execute
    self.call("execute")
  File "renpy/ast.py", line 2003, in call
    return renpy.statements.call(method, parsed, *args, **kwargs)
  File "renpy/statements.py", line 278, in call
    return method(parsed, *args, **kwargs)
  File "renpy/common/000statements.rpy", line 569, in execute_call_screen
    store._return = renpy.call_screen(name, *args, **kwargs)
  File "renpy/exports.py", line 3136, in call_screen
    rv = renpy.ui.interact(mouse="screen", type="screen", roll_forward=roll_forward)
  File "renpy/ui.py", line 292, in interact
    raise Exception("ui.interact called with non-empty widget/layer stack. Did you forget a ui.close() somewhere?\nStack was " + ('\n'.join([str(item) for item in stack])))
Exception: ui.interact called with non-empty widget/layer stack. Did you forget a ui.close() somewhere?
Stack was <Layer: u'transient'>
<renpy.ui.Detached object at 0x0000000007f27950>

Windows-10-10.0.19041
Ren'Py 7.4.11.2266
Demo 1.0
Fri Apr  1 19:37:01 2022
```

I can't reproduce these errors consistently, unfortunately. It seems to occur more frequently the closer I load to the time the save was made, but that might just be my own speculation. I also once got a 
```
  File "renpy/ui.py", line 191, in close
    raise Exception("Detached expects to be given a child.")
Exception: Detached expects to be given a child.
``` 
error, which points to the same `call screen loading` line as the other errors.

tl;dr, I keep getting weird error messages that seem to complain about being unable to load a screen somehow. These errors are not consistent, even when loading the same file multiple times, and come in four different flavours (all from loading the exact same file, before entering the game). Is this a problem on Ren'Py's end, or is there something I might be doing here to accidentally cause this?

I've left out further information on my game in particular so as not to make this exceedingly long in the hopes that these error messages on their own might point to a specific issue, but if I should provide more details, let me know.

**Comment by shawna-p** [comment]
I should also note - this happens exclusively when `config.developer = True`. I have never seen any of these errors on a distribution (which, at least, is nice).

**Comment by renpytom** [comment]
Hm...

**Comment by renpytom** [comment]
Can you check  renpy/renpy@3846046 , and see if that fixes it for you?  At first glance, this might deal with it - it's plausible, anyway.

Thanks for a fairly complete bug report, it pointed me in what I hope is the right direction.

**Comment by renpytom** [comment]
Closing this for now, please  re-open it if it's not fixed.
